In [25]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## **Import Dataset**

In [26]:
tr= r"C:\Users\fahru\Documents\SQL\Export Gcoffee\trx25.csv"
item= r"C:\Users\fahru\Documents\SQL\Export Gcoffee\item25.csv"
menu= r"C:\Users\fahru\Documents\SQL\database\Gcoffee\menu_items.csv"

trx_25= pd.read_csv(tr)
menus= pd.read_csv(menu)
item_25= pd.read_csv(item)

trx_25.head()

transaction_id  store_id  payment_method_id  \
0  00001472-e5a0-453f-9575-ecdfa6fe0e6c         9                  2   
1  00002670-7a56-485e-9240-0f9ec868b5c8         4                  4   
2  000028a4-1a9f-4e00-862d-4c69496a8015         5                  3   
3  00002a2b-e75d-4ce1-b0ad-b4c9dd962898         3                  4   
4  00003152-137d-4793-9430-2dc433eb22ad         8                  2   

   voucher_id    user_id  original_amount  discount_applied  final_amount  \
0         NaN  1929523.0             32.0               0.0          32.0   
1         NaN   267622.0             38.0               0.0          38.0   
2         NaN  1850723.0             19.0               0.0          19.0   
3         NaN  1366156.0             16.0               0.0          16.0   
4         NaN  1587555.0             63.0               0.0          63.0   

            created_at  
0  2025-05-17 10:17:23  
1  2025-04-01 15:21:07  
2  2025-05-03 15:51:29  
3  2025-05-21 14:24:00  
4  2025-05-16 13:15:20

In [27]:
item_25.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3640546 entries, 0 to 3640545
Data columns (total 6 columns):
 #   Column          Dtype  
---  ------          -----  
 0   transaction_id  object 
 1   item_id         int64  
 2   quantity        int64  
 3   unit_price      float64
 4   subtotal        float64
 5   created_at      object 
dtypes: float64(2), int64(2), object(2)
memory usage: 166.7+ MB


In [28]:
item_25.head()

transaction_id  item_id  quantity  unit_price  \
0  7ed3077f-2b58-44d0-8532-bbe247af63fe        8         1        10.0   
1  7ed3077f-2b58-44d0-8532-bbe247af63fe        5         3         9.0   
2  7ed3077f-2b58-44d0-8532-bbe247af63fe        4         1         8.0   
3  1f106ec5-25aa-4c85-97d7-2b90dfc30b3a        7         2         9.0   
4  d4f9df8d-03db-4dea-9370-00d37d3660f2        1         1         6.0   

   subtotal           created_at  
0      10.0  2025-04-01 09:59:32  
1      27.0  2025-04-01 09:59:32  
2       8.0  2025-04-01 09:59:32  
3      18.0  2025-04-01 09:59:34  
4       6.0  2025-04-01 09:59:35

In [29]:
menus= pd.read_csv(menu)
menus.head()

item_id   item_name category  price  is_seasonal  available_from  \
0        1    Espresso   coffee    6.0        False             NaN   
1        2   Americano   coffee    7.0        False             NaN   
2        3       Latte   coffee    8.0        False             NaN   
3        4  Cappuccino   coffee    8.0        False             NaN   
4        5  Flat White   coffee    9.0        False             NaN   

   available_to  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

In [30]:
trx6= trx_25[trx_25['store_id']==6].reset_index()
trx6.head()

index                        transaction_id  store_id  payment_method_id  \
0     16  000092da-9845-49bc-8ff6-fa2657b95261         6                  1   
1     25  0000c107-2d7d-4a83-a856-e68c14280df0         6                  3   
2     37  00010fc2-3dbc-4aff-9c8d-d735ee2bb93f         6                  3   
3     43  0001bb6b-12db-4da8-a66d-9cb43e280342         6                  4   
4     50  0001f12d-84bd-4279-8284-ac157369377c         6                  1   

   voucher_id    user_id  original_amount  discount_applied  final_amount  \
0         NaN  1158077.0             18.0               0.0          18.0   
1         NaN  2180411.0             21.5               0.0          21.5   
2         NaN  1995324.0             65.0               0.0          65.0   
3         NaN  1885869.0             48.0               0.0          48.0   
4         NaN  1890292.0             12.0               0.0          12.0   

            created_at  
0  2025-06-26 09:20:02  
1  2025-06-28 11:32:45  
2  2025-05-28 11:04:27  
3  2025-05-09 14:35:29  
4  2025-05-10 12:49:46

In [31]:
trx_detail= trx6[['transaction_id']].merge(item_25[['transaction_id','item_id','quantity','unit_price','subtotal']],
                       on='transaction_id', how='left')
trx_detail= trx_detail.merge(menus[['item_id','item_name','category','price']], on='item_id', how='left')
trx_detail.head()

transaction_id  item_id  quantity  unit_price  \
0  000092da-9845-49bc-8ff6-fa2657b95261        5         2         9.0   
1  0000c107-2d7d-4a83-a856-e68c14280df0        6         1         9.5   
2  0000c107-2d7d-4a83-a856-e68c14280df0        1         2         6.0   
3  00010fc2-3dbc-4aff-9c8d-d735ee2bb93f        5         3         9.0   
4  00010fc2-3dbc-4aff-9c8d-d735ee2bb93f        4         1         8.0   

   subtotal   item_name category  price  
0      18.0  Flat White   coffee    9.0  
1       9.5       Mocha   coffee    9.5  
2      12.0    Espresso   coffee    6.0  
3      27.0  Flat White   coffee    9.0  
4       8.0  Cappuccino   coffee    8.0

## **EDA**

In [32]:
trx_detail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 361528 entries, 0 to 361527
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   transaction_id  361528 non-null  object 
 1   item_id         361528 non-null  int64  
 2   quantity        361528 non-null  int64  
 3   unit_price      361528 non-null  float64
 4   subtotal        361528 non-null  float64
 5   item_name       361528 non-null  object 
 6   category        361528 non-null  object 
 7   price           361528 non-null  float64
dtypes: float64(3), int64(2), object(3)
memory usage: 22.1+ MB


In [33]:
menus

item_id      item_name    category  price  is_seasonal  available_from  \
0        1       Espresso      coffee    6.0        False             NaN   
1        2      Americano      coffee    7.0        False             NaN   
2        3          Latte      coffee    8.0        False             NaN   
3        4     Cappuccino      coffee    8.0        False             NaN   
4        5     Flat White      coffee    9.0        False             NaN   
5        6          Mocha      coffee    9.5        False             NaN   
6        7  Hot Chocolate  non-coffee    9.0        False             NaN   
7        8   Matcha Latte  non-coffee   10.0        False             NaN   

   available_to  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  
5           NaN  
6           NaN  
7           NaN

There are eight different menus that divided to two categories which are coffee and non-coffee product. The highest  product price is 'Matcha latte' and the lowest is 'Espresso'.

In [34]:
most= trx_detail.groupby('item_name').\
        agg(total_quantity=('quantity','sum'), total_revenue=('subtotal','sum')).reset_index()
most.sort_values('total_quantity', ascending=False)

item_name  total_quantity  total_revenue
6   Matcha Latte           91009       910090.0
5          Latte           90827       726616.0
4  Hot Chocolate           90571       815139.0
1     Cappuccino           90514       724112.0
2       Espresso           90448       542688.0
7          Mocha           90103       855978.5
0      Americano           89873       629111.0
3     Flat White           89381       804429.0

The best seller prodcut in store6 is 'Matcha Latte' and the worst is 'Flat Latte'. Mathca Latte is the most valueable product also.

## **Data Preparation**

In [35]:
# Create Dataframe 
basket= pd.pivot_table(data=trx_detail, index='transaction_id', columns='item_name',
                       values='item_id', aggfunc=pd.Series.nunique, fill_value=0)
basket

item_name                             Americano  Cappuccino  Espresso  \
transaction_id                                                          
000092da-9845-49bc-8ff6-fa2657b95261          0           0         0   
0000c107-2d7d-4a83-a856-e68c14280df0          0           0         1   
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f          0           1         0   
0001bb6b-12db-4da8-a66d-9cb43e280342          0           0         1   
0001f12d-84bd-4279-8284-ac157369377c          0           0         1   
...                                         ...         ...       ...   
ffff1a9c-a5f7-4d96-897f-8f8473d24d6d          0           0         0   
ffff5647-fe39-4426-8687-196fb458e68c          0           0         1   
ffff632b-e96e-4c8a-8e61-820fd5af9c2c          1           0         0   
ffff7567-7dc1-4fa9-8a72-6024cf19efd8          0           0         0   
ffff971e-ca8a-4863-85fd-b0deab016ccd          0           1         0   

item_name                             Flat White  Hot Chocolate  Latte  \
transaction_id                                                           
000092da-9845-49bc-8ff6-fa2657b95261           1              0      0   
0000c107-2d7d-4a83-a856-e68c14280df0           0              0      0   
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f           1              0      0   
0001bb6b-12db-4da8-a66d-9cb43e280342           0              0      0   
0001f12d-84bd-4279-8284-ac157369377c           0              0      0   
...                                          ...            ...    ...   
ffff1a9c-a5f7-4d96-897f-8f8473d24d6d           1              0      0   
ffff5647-fe39-4426-8687-196fb458e68c           0              0      1   
ffff632b-e96e-4c8a-8e61-820fd5af9c2c           0              0      0   
ffff7567-7dc1-4fa9-8a72-6024cf19efd8           0              0      1   
ffff971e-ca8a-4863-85fd-b0deab016ccd           0              1      1   

item_name                             Matcha Latte  Mocha  
transaction_id                                             
000092da-9845-49bc-8ff6-fa2657b95261             0      0  
0000c107-2d7d-4a83-a856-e68c14280df0             0      1  
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f             1      0  
0001bb6b-12db-4da8-a66d-9cb43e280342             1      0  
0001f12d-84bd-4279-8284-ac157369377c             0      0  
...                                            ...    ...  
ffff1a9c-a5f7-4d96-897f-8f8473d24d6d             0      0  
ffff5647-fe39-4426-8687-196fb458e68c             0      0  
ffff632b-e96e-4c8a-8e61-820fd5af9c2c             0      0  
ffff7567-7dc1-4fa9-8a72-6024cf19efd8             0      0  
ffff971e-ca8a-4863-85fd-b0deab016ccd             0      0  

[181128 rows x 8 columns]

In [36]:
# Check what menus ordered based on trnsacation_id
ro= basket.loc['ffff971e-ca8a-4863-85fd-b0deab016ccd']
nonzero= ro[ro>0].index.tolist()

basket.loc[['ffff971e-ca8a-4863-85fd-b0deab016ccd'],nonzero]

item_name                             Cappuccino  Hot Chocolate  Latte
transaction_id                                                        
ffff971e-ca8a-4863-85fd-b0deab016ccd           1              1      1

In [37]:
# Check item detail ordered based on transaction ID
trx_detail[trx_detail['transaction_id']=='ffff971e-ca8a-4863-85fd-b0deab016ccd']

transaction_id  item_id  quantity  unit_price  \
361525  ffff971e-ca8a-4863-85fd-b0deab016ccd        4         1         8.0   
361526  ffff971e-ca8a-4863-85fd-b0deab016ccd        7         3         9.0   
361527  ffff971e-ca8a-4863-85fd-b0deab016ccd        3         2         8.0   

        subtotal      item_name    category  price  
361525       8.0     Cappuccino      coffee    8.0  
361526      27.0  Hot Chocolate  non-coffee    9.0  
361527      16.0          Latte      coffee    8.0

In [38]:
# Encode Dataset
def encode(x):
    if x==0:
        return False
    if x>0:
        return True
basket= basket.map(encode)
basket

item_name                             Americano  Cappuccino  Espresso  \
transaction_id                                                          
000092da-9845-49bc-8ff6-fa2657b95261      False       False     False   
0000c107-2d7d-4a83-a856-e68c14280df0      False       False      True   
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f      False        True     False   
0001bb6b-12db-4da8-a66d-9cb43e280342      False       False      True   
0001f12d-84bd-4279-8284-ac157369377c      False       False      True   
...                                         ...         ...       ...   
ffff1a9c-a5f7-4d96-897f-8f8473d24d6d      False       False     False   
ffff5647-fe39-4426-8687-196fb458e68c      False       False      True   
ffff632b-e96e-4c8a-8e61-820fd5af9c2c       True       False     False   
ffff7567-7dc1-4fa9-8a72-6024cf19efd8      False       False     False   
ffff971e-ca8a-4863-85fd-b0deab016ccd      False        True     False   

item_name                             Flat White  Hot Chocolate  Latte  \
transaction_id                                                           
000092da-9845-49bc-8ff6-fa2657b95261        True          False  False   
0000c107-2d7d-4a83-a856-e68c14280df0       False          False  False   
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f        True          False  False   
0001bb6b-12db-4da8-a66d-9cb43e280342       False          False  False   
0001f12d-84bd-4279-8284-ac157369377c       False          False  False   
...                                          ...            ...    ...   
ffff1a9c-a5f7-4d96-897f-8f8473d24d6d        True          False  False   
ffff5647-fe39-4426-8687-196fb458e68c       False          False   True   
ffff632b-e96e-4c8a-8e61-820fd5af9c2c       False          False  False   
ffff7567-7dc1-4fa9-8a72-6024cf19efd8       False          False   True   
ffff971e-ca8a-4863-85fd-b0deab016ccd       False           True   True   

item_name                             Matcha Latte  Mocha  
transaction_id                                             
000092da-9845-49bc-8ff6-fa2657b95261         False  False  
0000c107-2d7d-4a83-a856-e68c14280df0         False   True  
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f          True  False  
0001bb6b-12db-4da8-a66d-9cb43e280342          True  False  
0001f12d-84bd-4279-8284-ac157369377c         False  False  
...                                            ...    ...  
ffff1a9c-a5f7-4d96-897f-8f8473d24d6d         False  False  
ffff5647-fe39-4426-8687-196fb458e68c         False  False  
ffff632b-e96e-4c8a-8e61-820fd5af9c2c         False  False  
ffff7567-7dc1-4fa9-8a72-6024cf19efd8         False  False  
ffff971e-ca8a-4863-85fd-b0deab016ccd         False  False  

[181128 rows x 8 columns]

In [39]:
# Take transaction that have two items or more
basket_filter= basket[(basket>0).sum(axis=1)>1]
basket_filter

item_name                             Americano  Cappuccino  Espresso  \
transaction_id                                                          
0000c107-2d7d-4a83-a856-e68c14280df0      False       False      True   
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f      False        True     False   
0001bb6b-12db-4da8-a66d-9cb43e280342      False       False      True   
000200e7-e08f-4e3d-a50b-20cafb90e209       True       False     False   
0002b217-3d63-4fad-9e39-c0b07e015918       True       False     False   
...                                         ...         ...       ...   
fffe675a-a67d-40ad-aa76-db195f20094f      False        True      True   
fffee8b8-af42-4db0-a694-b9f231059722       True       False     False   
fffef34b-f275-4c18-933e-9d1dd8639dae       True       False      True   
ffff5647-fe39-4426-8687-196fb458e68c      False       False      True   
ffff971e-ca8a-4863-85fd-b0deab016ccd      False        True     False   

item_name                             Flat White  Hot Chocolate  Latte  \
transaction_id                                                           
0000c107-2d7d-4a83-a856-e68c14280df0       False          False  False   
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f        True          False  False   
0001bb6b-12db-4da8-a66d-9cb43e280342       False          False  False   
000200e7-e08f-4e3d-a50b-20cafb90e209       False          False   True   
0002b217-3d63-4fad-9e39-c0b07e015918       False          False   True   
...                                          ...            ...    ...   
fffe675a-a67d-40ad-aa76-db195f20094f       False          False  False   
fffee8b8-af42-4db0-a694-b9f231059722       False           True  False   
fffef34b-f275-4c18-933e-9d1dd8639dae        True          False  False   
ffff5647-fe39-4426-8687-196fb458e68c       False          False   True   
ffff971e-ca8a-4863-85fd-b0deab016ccd       False           True   True   

item_name                             Matcha Latte  Mocha  
transaction_id                                             
0000c107-2d7d-4a83-a856-e68c14280df0         False   True  
00010fc2-3dbc-4aff-9c8d-d735ee2bb93f          True  False  
0001bb6b-12db-4da8-a66d-9cb43e280342          True  False  
000200e7-e08f-4e3d-a50b-20cafb90e209         False  False  
0002b217-3d63-4fad-9e39-c0b07e015918         False   True  
...                                            ...    ...  
fffe675a-a67d-40ad-aa76-db195f20094f         False   True  
fffee8b8-af42-4db0-a694-b9f231059722         False   True  
fffef34b-f275-4c18-933e-9d1dd8639dae         False  False  
ffff5647-fe39-4426-8687-196fb458e68c         False  False  
ffff971e-ca8a-4863-85fd-b0deab016ccd         False  False  

[111999 rows x 8 columns]

In [40]:
# Check

basket_filter.loc[['0002b217-3d63-4fad-9e39-c0b07e015918']]

item_name                             Americano  Cappuccino  Espresso  \
transaction_id                                                          
0002b217-3d63-4fad-9e39-c0b07e015918       True       False     False   

item_name                             Flat White  Hot Chocolate  Latte  \
transaction_id                                                           
0002b217-3d63-4fad-9e39-c0b07e015918       False          False   True   

item_name                             Matcha Latte  Mocha  
transaction_id                                             
0002b217-3d63-4fad-9e39-c0b07e015918         False   True

## **Analysis**

In [41]:
# USe apriori algorithm for analysis
from mlxtend.frequent_patterns import apriori

freq_item= apriori(basket_filter, min_support=.01, use_colnames=True).\
            sort_values('support', ascending=False).reset_index(drop=True)
freq_item['product_count']= freq_item['itemsets'].apply(lambda x: len(x))
freq_item

support                       itemsets  product_count
0   0.295744                 (Matcha Latte)              1
1   0.295458                        (Latte)              1
2   0.295119                   (Cappuccino)              1
3   0.295029                (Hot Chocolate)              1
4   0.293253                   (Flat White)              1
5   0.293137                    (Americano)              1
6   0.292101                     (Espresso)              1
7   0.291279                        (Mocha)              1
8   0.061840          (Latte, Matcha Latte)              2
9   0.061786     (Matcha Latte, Cappuccino)              2
10  0.061402         (Latte, Hot Chocolate)              2
11  0.061393              (Latte, Espresso)              2
12  0.061376      (Matcha Latte, Americano)              2
13  0.061358      (Hot Chocolate, Espresso)              2
14  0.061340         (Mocha, Hot Chocolate)              2
15  0.061331  (Hot Chocolate, Matcha Latte)              2
16  0.061331                 (Mocha, Latte)              2
17  0.061268            (Flat White, Latte)              2
18  0.061242        (Flat White, Americano)              2
19  0.061224        (Cappuccino, Americano)              2
20  0.061224    (Hot Chocolate, Cappuccino)              2
21  0.061215     (Hot Chocolate, Americano)              2
22  0.060956            (Latte, Cappuccino)              2
23  0.060840       (Flat White, Cappuccino)              2
24  0.060831     (Flat White, Matcha Latte)              2
25  0.060563         (Flat White, Espresso)              2
26  0.060527       (Espresso, Matcha Latte)              2
27  0.060465            (Mocha, Cappuccino)              2
28  0.060242    (Flat White, Hot Chocolate)              2
29  0.060099          (Espresso, Americano)              2
30  0.060090            (Mocha, Flat White)              2
31  0.060054             (Latte, Americano)              2
32  0.059822          (Mocha, Matcha Latte)              2
33  0.059670             (Mocha, Americano)              2
34  0.059384         (Espresso, Cappuccino)              2
35  0.059367              (Mocha, Espresso)              2

In [42]:
# Summary variabels

from mlxtend.frequent_patterns import association_rules

product_assc= association_rules(freq_item, metric='confidence', min_threshold=.21).\
                sort_values(['support','confidence'],ascending=[False,False]).reset_index(drop=True)
product_assc

antecedents      consequents  antecedent support  consequent support  \
0  (Espresso)          (Latte)            0.292101            0.295458   
1  (Espresso)  (Hot Chocolate)            0.292101            0.295029   
2     (Mocha)  (Hot Chocolate)            0.291279            0.295029   
3     (Mocha)          (Latte)            0.291279            0.295458   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.061393    0.210179  0.711366               1.0 -0.024910    0.892027   
1  0.061358    0.210057  0.711985               1.0 -0.024821    0.892432   
2  0.061340    0.210588  0.713785               1.0 -0.024596    0.893032   
3  0.061331    0.210557  0.712646               1.0 -0.024730    0.892455   

   zhangs_metric   jaccard  certainty  kulczynski  
0      -0.364340  0.116681  -0.121042    0.208985  
1      -0.363642  0.116700  -0.120534    0.209014  
2      -0.361342  0.116845  -0.119781    0.209249  
3      -0.362628  0.116730  -0.120505    0.209068

In [43]:
product_assc.sort_values('antecedent support', ascending=False)

antecedents      consequents  antecedent support  consequent support  \
0  (Espresso)          (Latte)            0.292101            0.295458   
1  (Espresso)  (Hot Chocolate)            0.292101            0.295029   
2     (Mocha)  (Hot Chocolate)            0.291279            0.295029   
3     (Mocha)          (Latte)            0.291279            0.295458   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.061393    0.210179  0.711366               1.0 -0.024910    0.892027   
1  0.061358    0.210057  0.711985               1.0 -0.024821    0.892432   
2  0.061340    0.210588  0.713785               1.0 -0.024596    0.893032   
3  0.061331    0.210557  0.712646               1.0 -0.024730    0.892455   

   zhangs_metric   jaccard  certainty  kulczynski  
0      -0.364340  0.116681  -0.121042    0.208985  
1      -0.363642  0.116700  -0.120534    0.209014  
2      -0.361342  0.116845  -0.119781    0.209249  
3      -0.362628  0.116730  -0.120505    0.209068

## **Result**

Espresso has ordered 5.9% percentage from total transaction. After buy a the Espresso, cust will ordered latte too with 21% probability. Espresso and latte have negative correlation(n<1), that means the probability of these product bought together lower than these product bought for itself.

FOr this case, there are no one combination that realy have big correlation. That can be indicated by the highest confidence value is 0.21 which means there are no combination have probability higher thatn 21% to bought them together. The most cust seems to order just one menu for each transaction. 